lib/puppet/configurer.rb

```diff
@@ -181,25 +181,50 @@ def apply_catalog(catalog, options)
   # which accepts :tags and :ignoreschedules.
   def run(options = {})
     pool = Puppet::Network::HTTP::Pool.new(Puppet[:http_keepalive_timeout])
+    # We create the report pre-populated with default settings for
+    # environment and transaction_uuid very early, this is to ensure
+    # they are sent regardless of any catalog compilation failures or
+    # exceptions.
+    options[:report] ||= Puppet::Transaction::Report.new("apply", nil, @environment, @transaction_uuid)
+    report = options[:report]
+    init_storage
+
+    Puppet::Util::Log.newdestination(report)
+
     begin
       Puppet.override(:http_pool => pool) do
-        run_internal(options)
+
+        # Skip failover logic if the server_list setting is empty
+        if Puppet.settings[:server_list].nil? || Puppet.settings[:server_list].empty?
+          do_failover = false;
+        else
+          do_failover = true
+        end
+        # When we are passed a catalog, that means we're in apply
+        # mode. We shouldn't try to do any failover in that case.
+        if options[:catalog].nil? && do_failover
+          server = find_functional_server()
+          if server.nil?
+            Puppet.warning "Could not select a functional puppet master"
+            server = [nil, nil]
+          end
+          server[1] ||= Puppet.settings[:masterport]
+          Puppet.override(:server => server[0], :serverport => server[1]) do
+            Puppet.debug "Selected master: #{server[0]}:#{server[1]}"
+
+            run_internal(options)
+          end
+        else
+          run_internal(options)
+        end
       end
     ensure
       pool.close
     end
   end
 
   def run_internal(options)
-    # We create the report pre-populated with default settings for
-    # environment and transaction_uuid very early, this is to ensure
-    # they are sent regardless of any catalog compilation failures or
-    # exceptions.
-    options[:report] ||= Puppet::Transaction::Report.new("apply", nil, @environment, @transaction_uuid)
     report = options[:report]
```

Comment:
This is used. I see we're not consistent about using `report` vs `options[:report]` later on.